electron/libs/cli.ts

```diff
@@ -50,76 +53,94 @@ function newInstallProgressNotifier(full_name: string, notifyMainWindow: MainWin
+  const cli = await installTeaCli();
+  const project = guipkg.full_name; //TODO make the naming in the API consistent with libtea and tea/cli thanks
 
-  const scriptPath = await createCommandScriptFile(cmd);
-  try {
-    let stdout = "";
-    let stderr = "";
+  if (project in ptys) {
+    /// already running!
+    return;
+  }
 
-    await new Promise((resolve, reject) => {
-      const child = spawn("/usr/bin/osascript", [scriptPath]);
-      child.stdout.on("data", (data) => {
-        stdout += data.toString().trim();
-      });
-      child.stderr.on("data", (data) => {
-        stderr += data.toString().trim();
-      });
+  // look up the entrypoint for the package again in case it changed. This makes
+  // hacking on the entrypoint more ergonomic for local, third party developers
+  let { entrypoint } = await getPantryDetails(project);
+  entrypoint ??= `tea +${project} sh`;
+
+  notifyMainWindow("pty.out", { project, type: "reset" });
+
+  const installation = await tea.hooks
+    .useCellar()
+    .resolve({ project, constraint: new tea.semver.Range("*") });
+  const shell =
+    process.platform === "win32" ? "powershell.exe" : process.platform == "darwin" ? "zsh" : "bash";
+  const wet = await tea.plumbing.hydrate(installation.pkg);
+  const gas = await tea.plumbing.resolve(wet.pkgs);
+  if (gas.pending.length) throw new Error("pkg dependencies not installed");
+  const cwd = installation.path.string;
+
+  const installations = gas.installed;
+  if (!installations.some(({ pkg: { project } }) => project == "tea.xyz")) {
+    // dont’t dupe tea/cli if it’s already a dep
+    // ∵ useShellEnv is plumbing and thus will happily dupe env
+    installations.push(cli);
+  }
 
-      child.on("exit", (code) => {
-        log.info("exit:", code, `\`${stdout}\``);
-        if (code == 0) {
-          resolve(stdout);
-        } else {
-          reject(new Error("failed to open terminal and run tea sh"));
-        }
-      });
+  const rawenv = await tea.hooks.useShellEnv().map({ installations });
+  const env = tea.hooks.useShellEnv().flatten(rawenv);
+  env.PATH += ":/usr/bin:/bin";
+  env.TMPDIR ??= process.env.TMPDIR ?? "error";
+  env.HOME ??= process.env.HOME ?? "error";
+  env.TEA_GUI = "1";
+
+  /// user env is likely to screw up our env configuration
+  const args = ["--no-rcs"];
+
+  const ptyproc = pty.spawn(shell, args, {
+    name: "xterm-color", //NOTE just copied this from the docs so dunno if it’s a good choice or what
+    cwd, //TODO probs not a good CWD choice, probs make a new dir for each project in our tmpdir
+    env
+  });
+
+  ptyproc.onData((out) =>
+    notifyMainWindow("pty.out", {
+      project,
+      out,
+      type: "out"
+    })
+  );
+  ptyproc.onExit(({ exitCode }) =>
+    notifyMainWindow("pty.out", {
+      project,
+      out: `exited with ${exitCode}`,
+      type: "out"
+    })
+  );
+
+  ptyproc.write(`${entrypoint}\r`);
+
+  ipcMain.handle("pty.in", (event, { data, project: incoming_project }) => {
```

Comment:
this seems wrong, as in: we never unsubscribe. didn't read the docs so dunno.

Reply:
i wonder if you can break it. start and end the same pty a bunch of times to get multiple registered handlers, maybe?

Reply:
It is wrong.  I got it to crash for me, so I read the docs :-)